apps/meteor/client/components/message/toolbar/MessageToolbarStarsActionMenu.tsx

```diff
@@ -0,0 +1,97 @@
+import { useUniqueId } from '@rocket.chat/fuselage-hooks';
+import { GenericMenu, type GenericMenuItemProps } from '@rocket.chat/ui-client';
+import { useTranslation } from '@rocket.chat/ui-contexts';
+import type { MouseEvent, ReactElement } from 'react';
+import React from 'react';
+
+import type { MessageActionConditionProps, MessageActionConfig } from '../../../../app/ui-utils/client/lib/MessageAction';
+
+type MessageActionConfigOption = Omit<MessageActionConfig, 'condition' | 'context' | 'order' | 'action'> & {
+	action: (e?: MouseEvent<HTMLElement>) => void;
+};
+
+type MessageActionSection = {
+	id: string;
+	title: string;
+	items: GenericMenuItemProps[];
+};
+
+type MessageActionMenuProps = {
+	onChangeMenuVisibility: (visible: boolean) => void;
+	options: MessageActionConfigOption[];
+	context: MessageActionConditionProps;
+	isMessageEncrypted: boolean;
+};
+
+const MessageToolbarStarsActionMenu = ({
+	options,
+	onChangeMenuVisibility,
+	context,
+	isMessageEncrypted,
+}: MessageActionMenuProps): ReactElement => {
+	const t = useTranslation();
+	const id = useUniqueId();
+	const groupOptions = options
+		.map((option) => ({
```

Comment:
Why map, reduce, map while you can do the whole in a single reduce?

Reply:
We possibly can, but I didn't want to modify this logic here since another function used the same and would have been a good addition as a separate PR. But if you feel like it's good to have here, I don't mind.